### PR TITLE
Add support for multiple fire animations in WithInfantryBody 

### LINKED ISSUE #11892: WithInfantryBody only supports a single firing animation.
This is a problem with artwork like http://ppmforums.com/viewtopic.php?t=42518 which has more firing animations for different weapons.

### PR DESCRIPTION
**Pull request by abc013**
Fixes #11892.
Also changed the name of the secondary weapon of the Rocket Soldier because it had two primary armaments.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Render/WithInfantryBody.cs [hunk 1]
```diff
@@ -25,9 +25,21 @@ public class WithInfantryBodyInfo : UpgradableTraitInfo, IRenderActorPreviewSpri
 
 		[SequenceReference] public readonly string MoveSequence = "run";
 		[SequenceReference] public readonly string AttackSequence = null;
+
+		[FieldLoader.LoadUsing("LoadWeaponSequences")]
+		[SequenceReference] public readonly Dictionary<string, string> AttackSequenceByArmament = new Dictionary<string, string>();
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Render/WithInfantryBody.cs]: My `[SequenceReference]` or rather the accompanying lint check isn't smart enough for dictionaries. Remove it and leave a `TODO` as I assume this won't be easy to solve.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithInfantryBody.cs]: The name is a bit overboard IMO, `AttackSequences` with a desc explaining that it is a dictionary of armament names to animation would be fine.

**Comment by abc013** [comment]
Updated.

**Comment by abc013** [comment]
Updated again because i forgot to write the `//TODO`.

**Comment by Phrohdoh** [comment]
This should remove `AttackSequence` and include an upgrade rule or rename `AttackSequence` to `DefaultAttackSequence`.

File: OpenRA.Mods.Common/Traits/Render/WithInfantryBody.cs [hunk 2]
```diff
@@ -25,9 +25,23 @@ public class WithInfantryBodyInfo : UpgradableTraitInfo, IRenderActorPreviewSpri
 
 		[SequenceReference] public readonly string MoveSequence = "run";
 		[SequenceReference] public readonly string AttackSequence = null;
+
+		// TODO: [SequenceReference] isn't smart enough to use Dictionarys.
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Render/WithInfantryBody.cs]: Nit: Dictionaries.

**Comment by reaperrr** [comment]
> This should remove AttackSequence and include an upgrade rule or rename AttackSequence to DefaultAttackSequence.

I agree, and while I'm aware it'll be more work, I think removing `AttackSequence` and including an upgrade rule would be the more straightforward solution (having two properties would be a little confusing, even with the rename).

**Comment by pchote** [comment]
Forcing all mods to define sequences for all their armaments kind of sucks.

**Comment by reaperrr** [comment]
> Forcing all mods to define dictionary entries for all their armaments kind of sucks.

Hm, good point. I guess the rename to `DefaultAttackSequence` will do.

**Comment by abc013** [comment]
Ok, I renamed `AttackSequence` to `DefaultAttackSequence`.

File: OpenRA.Mods.Common/Traits/Render/WithInfantryBody.cs [hunk 3]
```diff
@@ -24,10 +24,24 @@ public class WithInfantryBodyInfo : UpgradableTraitInfo, IRenderActorPreviewSpri
 		public readonly int MaxIdleDelay = 110;
 
 		[SequenceReference] public readonly string MoveSequence = "run";
-		[SequenceReference] public readonly string AttackSequence = null;
+		[SequenceReference] public readonly string DefaultAttackSequence = null;
+
+		// TODO: [SequenceReference] isn't smart enough to use Dictionarys.
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Render/WithInfantryBody.cs]: Nit: Dictionaries.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 4]
```diff
@@ -386,6 +386,17 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
 					}
 				}
 
+				// Renamed AttackSequence to DefaultAttackSequence in WithInfantryBody.
+				if (engineVersion < 20160908)
```
Comment (reaperrr) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: I think you can update this to something like 20161016 for now, maybe we'll get it merged quickly.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
Works as advertised, :+1: after the fixups.

**Comment by abc013** [comment]
Updated.

**Comment by abcdefg30** [comment]
👍 / ✅

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/19ad799d70fa51c8446936fbd43415222b8a771d...f36dc2eeebfc6829445edb45bd7c07c38b9bec7d)

File: OpenRA.Mods.Common/Traits/Render/WithInfantryBody.cs [hunk 5]
```diff
@@ -96,18 +110,22 @@ protected virtual bool AllowIdleAnimation(Actor self)
...
 				state = AnimationState.Attacking;
-				DefaultAnimation.PlayThen(NormalizeInfantrySequence(self, Info.AttackSequence), () => state = AnimationState.Idle);
+				DefaultAnimation.PlayThen(NormalizeInfantrySequence(self, sequence), () => state = AnimationState.Idle);
 			}
 		}
 
 		void INotifyAttack.PreparingAttack(Actor self, Target target, Armament a, Barrel barrel)
 		{
-			Attacking(self, target);
+			Attacking(self, target, a);
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Render/WithInfantryBody.cs]: I believe this change triggered a Coverity defect: https://scan8.coverity.com/reports.htm#v24414/p10888/fileInstanceId=13580582&defectInstanceId=4704498&mergedDefectId=153439

**Pull request merged**